File: onedesk/one_hr/learned.py
"""Settling something the clock proposed.

`healing.networks` and `healing.zones` write a `Clock Network` or a `Clock
Place` as Proposed on the first corroborated sighting and promote it to
Confirmed once enough employees agree. The two states a person decides between
are the same on both doctypes, so the verbs are here rather than twice.

Status is read only on those records. It was a dropdown, which made rejecting
a proposal look like picking a value from a list, and let somebody set a
rejected address back to Confirmed without saying why. A rejection is final in
the sense that matters: `healing._learn` matches a Rejected row like any other
and never promotes it again.
"""

import frappe
from frappe import _

#: The doctypes the learner writes. Nothing else may be settled through here.
LEARNED = ("Clock Network", "Clock Place")


def _settle(doctype: str, name: str, status: str) -> None:
	if doctype not in LEARNED:
		frappe.throw(_("{0} is not something the clock proposes.").format(doctype))
	frappe.has_permission(doctype, "write", doc=name, throw=True)

	frappe.db.set_value(doctype, name, "status", status)
	frappe.get_doc(doctype, name).add_comment("Comment", _("Marked {0}.").format(_(status)))


@frappe.whitelist(methods=["POST"])
def confirm(doctype: str, name: str) -> None:
	"""Allow check-ins from it, ahead of the threshold agreeing."""
	_settle(doctype, name, "Confirmed")


@frappe.whitelist(methods=["POST"])
def reject(doctype: str, name: str) -> None:
	"""Refuse it, and stop the learner offering it again."""
	_settle(doctype, name, "Rejected")

File: onedesk/public/js/learned.js
// The two decisions on something the clock proposed. Status is read-only on
// the record, because rejecting an address is a decision rather than a value
// you pick out of a list — and because the learner treats a rejection as final.
frappe.ui.form.on("Clock Network", { refresh: settle });

function settle(frm) {
	if (frm.is_new() || !frappe.model.can_write(frm.doctype)) return;

	if (frm.doc.status === "Proposed") {
		frm.add_custom_button(__("Confirm"), () =>
			ask(frm, "confirm", __("Check-ins from {0} will be allowed.", [frm.doc.address || frm.doc.label])),
		);
	}

	if (["Proposed", "Confirmed"].includes(frm.doc.status)) {
		frm.add_custom_button(__("Reject"), () =>
			ask(
				frm,
				"reject",
				__("Check-ins from {0} will be refused, and it will not be proposed again.", [
					frm.doc.address || frm.doc.label,
				]),
			),
		);
	}
}

function ask(frm, verb, question) {
	frappe.confirm(question, () =>
		frappe
			.xcall(`onedesk.one_hr.learned.${verb}`, { doctype: frm.doctype, name: frm.doc.name })
			.then(() => {
				frappe.show_alert({ message: __("Done"), indicator: "green" });
				frm.reload_doc();
			}),
	);
}
